src/components/ContactsAdd.js

```diff
@@ -1,32 +1,140 @@
+    e.preventDefault()
+    if (addressData.id) {
+      fetch(`http://localhost:4000/contacts/${addressData.id}`, {
+        method: "PATCH",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify(addressData),
+      })
+        .then((res) => res.json())
+        .then((json) => {
+          const editedAddress = contacts.map(contact =>
+            contact.id === json.id ? json : contact
+          )
+          setContacts(editedAddress)
+          navigate("/")
+        })
+    } else {
 
-  //TODO: Implement controlled form
-  //send POST to json server on form submit
+      // Here the fetch is called if the ID field is empty (and therefore the
+      // rest of them will be too) and it POSTS the contents of the form 
+      // to the server and sets the value of contacts to a copy of the old array
+      // with the new data added
+      fetch("http://localhost:4000/contacts", {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify(addressData),
+      })
+        .then((res) => res.json())
+        .then((json) => {
+          console.log("Address created!", json)
+          setContacts([...contacts, json])
+          navigate(`/`)
+        })
+    }
+  }
+      // I trimmed a load out of here by replacing the if statements
+      // with the e.target.name key in [] so that this is dynamic
+  const handleChange = (e) => {
+    setAddressData({ ...addressData, [e.target.name]: e.target.value })
+  }
 
   return (
-    <form className="form-stack contact-form">
+    <form className="form-stack contact-form" onSubmit={handleSubmit}>
       <h2>Create Contact</h2>
-
       <label htmlFor="firstName">First Name</label>
-      <input id="firstName" name="firstName" type="text" required />
-
+      <input
+        id="firstName"
+        value={addressData.firstName}
+        name="firstName"
+        type="text"
+        onChange={handleChange}
+      />
       <label htmlFor="lastName">Last Name:</label>
-      <input id="lastName" name="lastName" type="text" required/>
-
+      <input
+        id="lastName"
+        value={addressData.lastName}
+        name="lastName"
+        type="text"
+        onChange={handleChange}
+      />
       <label htmlFor="street">Street:</label>
-      <input id="street" name="street" type="text" required/>
-
+      <input
+        id="street"
+        value={addressData.street}
+        name="street"
+        type="text"
+        onChange={handleChange}
+      />
       <label htmlFor="city">City:</label>
-      <input id="city" name="city" type="text" required/>
-
+      <input
+        id="city"
+        value={addressData.city}
+        name="city"
+        type="text"
+        onChange={handleChange}
+      />
+      <label htmlFor="email">Email:</label>
+      <input
+        id="email"
+        value={addressData.email}
+        name="email"
+        type="text"
```

Comment:
As a suggestion, you could try using type="email" here for some validation on the input.